nixos/modules/services/monitoring/sensu/default.nix

```diff
@@ -0,0 +1,476 @@
+{ config, pkgs, lib, ... }:
+
+let
+  cfg = config.services.sensu;
+  pgm = cfg.programs;
+
+  baseDir = "/etc/sensu";
+
+  transportIsRabbitMQ = (cfg.transport.name == "rabbitmq");
+  roleIsServer = (cfg.role == "server");
+
+  # setting the LOG_LEVEL env var doesn't work, so we force the logLevel via -L
+  bin = binary:
+    "${cfg.package}/bin/sensu-${binary} -L ${cfg.logLevel} -c ${baseDir}/config.json -d ${baseDir}/conf.d";
+
+  clientDeps = with pkgs; [ gawk ]
+    ++ lib.optionals pgm.common  (with pkgs; [ nagiosPluginsOfficial ])
+    ++ lib.optionals pgm.esx     (with pkgs; [ check-esx-hardware ])
+    ++ lib.optionals pgm.http    (with pkgs; [ curl ])
+    ++ lib.optionals pgm.ipmi    (with pkgs; [ ipmitool which ])
+    ++ lib.optionals pgm.openvpn (with pkgs; [ sensu-check-openvpn ])
+    ++ lib.optionals pgm.snmp    (with pkgs; [ net_snmp ]);
+
+  sensuService = desc: bin: let
+    pgm = lib.toLower desc;
+    isClient = lib.elem pgm [ "client" ];
+    isDash   = lib.elem pgm [ "dashboard" ];
+    isServer = lib.elem pgm [ "server" "api" ];
+  in {
+    description = "Sensu ${desc}";
+    wantedBy    = [ "sensu.target" ];
+    after       = [] ++ lib.optionals isServer [ "redis.service" "rabbmitmq.service" ];
+    path        = [ pkgs.bash cfg.package ] ++ cfg.dependencies ++ lib.optionals isClient clientDeps;
+    # needed for bundler
+    environment = {
+      HOME = "/tmp";
+    };
+    restartTriggers = [ ]
+      ++ lib.optionals isClient [ sensuCfg clientCfg ]
+      ++ lib.optionals isDash   [ dashCfg ]
+      ++ lib.optionals isServer [ sensuCfg checkCfg handlerCfg ];
+    serviceConfig = {
+      ExecStart        = bin;
+      DynamicUser      = true;
+      Slice            = "sensu.slice";
+      Restart          = "on-failure";
+      RestartSec       = "2s";
+      ProtectSystem    = "full";
+      ProtectHome      = true;
+      # sensu expects a common file in /tmp which must be shared with the other sensu units
+      # this can be enabled when our systemd module gets support for JoinsNamespaceOf.
+      PrivateTmp       = false;
+    };
+  };
+
+  cleanAttrSet = set_or_list:
+    if builtins.isList set_or_list
+      then (map (s: cleanAttrSet s) set_or_list)
+      else
+        lib.mapAttrsRecursive (path: v:
+          (if builtins.isString v then (lib.replaceStrings ["\n"] [""] v) else v))
+          (lib.filterAttrsRecursive (k: v:
+            !(k == "_module" || v == "" || v == [] || v == {})) set_or_list);
+
+  checkCfg = pkgs.writeText "checks.json" (builtins.toJSON (cleanAttrSet {
+    inherit (cfg) checks;
+  }));
```

Comment:
There are quite a few options, you have added here. Do you think you can maintain it in future versions that or would it be better to relax the checks and just convert attributes to JSON.